composer/lib/dependabot/composer/metadata_finder.rb

```diff
@@ -28,27 +28,26 @@ def look_up_source_from_packagist
         return nil if packagist_listing&.fetch("packages", nil) == []
         return nil unless packagist_listing&.dig("packages", dependency.name.downcase)
 
-        version_listings =
-          packagist_listing["packages"][dependency.name.downcase].
-          select { |version, _| Composer::Version.correct?(version) }.
-          sort_by { |version, _| Composer::Version.new(version) }.
-          map { |_, listing| listing }.
-          reverse
-
-        potential_source_urls =
-          version_listings.
-          flat_map { |info| [info["homepage"], info.dig("source", "url")] }.
-          compact
-
-        source_url = potential_source_urls.find { |url| Source.from_url(url) }
-
-        Source.from_url(source_url)
+        version_listings = packagist_listing["packages"][dependency.name.downcase]
+        # Packagist returns an array of version listings sorted newest to oldest.
+        # So iterate until we find the first URL that appears to be a source URL.
+        #
+        # NOTE: Each listing may not have all fields because they are minified to remove duplicate elements:
+        # * https://github.com/composer/composer/blob/main/UPGRADE-2.0.md#for-composer-repository-implementors
+        # * https://github.com/composer/metadata-minifier
+        version_listings.each do |i|
```

Comment:
Instead of using two nested loops, we could use a single loop to iterate through all the URLs in the version_listings array. This will eliminate the need to loop through the homepage and source.url keys separately, and it will also allow you to return the first non-nil source_url value. 

```ruby
version_listings.each do |i|
  url = i["homepage"] || i.dig("source", "url")
  source_url = Source.from_url(url)
  return source_url unless source_url.nil?
end


```

Reply:
The code is effectively identical from a perf standpoint... Because both are iterating through the exact same list of things, and then breaking on the first non-nil match. 

The only difference I can think of is if Ruby does something funky under our feet at the compiler level of setting up a loop instead of an `||` statement... but I doubt it.

I'll update to this though, as it strikes me as a little more readable... 👍

Reply:
Regardless form a perf perspective, the nested for-loop isn't a bad thing because it will _immediately_ `return` on the first match, so even if we flatten the loop (which is what ☝️ is suggesting) then it's still the same runtime complexity... because we still have to evaluate the same number of things.

Reply:
The below should fix the bug. But your logic seems to be fine as well. Since it is tested. I will go with it. 

```ruby
source_url = version_listings.find do |i|
  url = i["homepage"] || i.dig("source", "url")
  source = Source.from_url(url)
  !source.nil?
end
return source_url unless source_url.nil?


```